src/main/java/ratpack/zipkin/internal/DefaultServerTracingHandler.java

```diff
@@ -7,62 +7,103 @@
 import brave.http.HttpTracing;
 import brave.propagation.TraceContext;
 import javax.inject.Inject;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 import ratpack.handling.Context;
 import ratpack.handling.Handler;
+import ratpack.http.Headers;
+import ratpack.http.HttpMethod;
 import ratpack.http.Request;
 import ratpack.http.Response;
+import ratpack.http.Status;
+import ratpack.path.PathBinding;
+import ratpack.zipkin.ServerRequest;
+import ratpack.zipkin.ServerResponse;
 import ratpack.zipkin.ServerTracingHandler;
+import ratpack.zipkin.SpanNameProvider;
 
 /**
  * {@link Handler} for Zipkin tracing.
  */
 public final class DefaultServerTracingHandler implements ServerTracingHandler {
 
   private final Tracing tracing;
-  private final HttpServerHandler<Request, Response> handler;
-  private final TraceContext.Extractor<Request> extractor;
-
+  private final HttpServerHandler<ServerRequest, ServerResponse> handler;
+  private final TraceContext.Extractor<ServerRequest> extractor;
+  private final SpanNameProvider spanNameProvider;
+  private final Logger logger = LoggerFactory.getLogger(DefaultServerTracingHandler.class);
   @Inject
-  public DefaultServerTracingHandler(HttpTracing httpTracing) {
+  public DefaultServerTracingHandler(final HttpTracing httpTracing, final SpanNameProvider spanNameProvider) {
     this.tracing = httpTracing.tracing();
-    this.handler = HttpServerHandler.create(httpTracing, new HttpAdapter());
-    this.extractor = tracing.propagation().extractor((Request r, String name) -> r.getHeaders().get(name));
+    this.handler = HttpServerHandler.<ServerRequest, ServerResponse>create(httpTracing, new ServerHttpAdapter());
+    this.extractor = tracing.propagation().extractor((ServerRequest r, String name) -> r.getHeaders().get(name));
+    this.spanNameProvider = spanNameProvider;
   }
 
   @Override
   public void handle(Context ctx) throws Exception {
-    Request request = ctx.getRequest();
+    ServerRequest request = new ServerRequestImpl(ctx.getRequest());
     final Span span = handler.handleReceive(extractor, request);
     final Tracer.SpanInScope scope = tracing.tracer().withSpanInScope(span);
+
     ctx.getResponse().beforeSend(response -> {
-      handler.handleSend(response, null, span);
+      ServerResponseImpl wrappedResponse = new ServerResponseImpl(response, ctx);
+      String name = spanNameProvider.spanName(request, wrappedResponse);
+      span.name(name);
+      handler.handleSend(wrappedResponse, null, span);
       span.finish();
       scope.close();
     });
     span.start();
     ctx.next();
   }
 
-  static final class HttpAdapter extends brave.http.HttpServerAdapter<Request, Response> {
+  private static class ServerRequestImpl implements ServerRequest {
```

Comment:
This diff is totally messed up, but the adapter has been moved out. Just felt a little tidier.